### ISSUE TITLE
A lot of functions do not return the correct types or take incorrect parameters

### ISSUE THREAD

**Issue by melody-rs** [lebedec/libfmod#6]
I noticed that a lot of functions are returning types or results they should not be-

For example: All `is_valid` functions return `FMOD_BOOL` (a typedef for `c_int`) which `libfmod` interprets as `Result<(), libfmod::Error>`, rather than a `bool`.

Other functions, like `libfmod::Studio::load_bank_memory` take a `String` when they really should take bytes. Functions like these are expecting a `char*` *as* bytes, not a string. This relegates these functions to be practically useless.
```rs
    pub fn load_bank_memory(
        &self,
        buffer: &str,
        length: i32,
        mode: LoadMemoryMode,
        flags: ffi::FMOD_STUDIO_LOAD_BANK_FLAGS,
    ) -> Result<Bank, Error> {
        unsafe {
            let mut bank = null_mut();
            match ffi::FMOD_Studio_System_LoadBankMemory(
                self.pointer,
                CString::new(buffer)?.as_ptr(),
                length,
                mode.into(),
                flags,
                &mut bank,
            ) {
                ffi::FMOD_OK => Ok(Bank::from(bank)),
                error => Err(err_fmod!("FMOD_Studio_System_LoadBankMemory", error)),
            }
        }
    }
```
Should be:
```rs
    pub fn load_bank_memory(
        &self,
        buffer: Vec<u8>,
        mode: LoadMemoryMode,
        flags: ffi::FMOD_STUDIO_LOAD_BANK_FLAGS,
    ) -> Result<Bank, Error> {
        unsafe {
            let mut bank = null_mut();
            match ffi::FMOD_Studio_System_LoadBankMemory(
                self.pointer,
                CString::from_vec_unchecked(buffer).as_ptr(),
                buffer.len(),
                mode.into(),
                flags,
                &mut bank,
            ) {
                ffi::FMOD_OK => Ok(Bank::from(bank)),
                error => Err(err_fmod!("FMOD_Studio_System_LoadBankMemory", error)),
            }
        }
    }
```

There's also some functions like `libfmod::Bank::get_path` which take in the length of a String? and do nothing with it which makes no sense.
```rs
    pub fn get_path(&self, size: i32) -> Result<(String, i32), Error> {
        unsafe {
            let path = CString::from_vec_unchecked(b"".to_vec()).into_raw();
            let mut retrieved = i32::default();
            match ffi::FMOD_Studio_Bank_GetPath(self.pointer, path, size, &mut retrieved) {
                ffi::FMOD_OK => Ok((
                    CString::from_raw(path)
                        .into_string()
                        .map_err(Error::String)?,
                    retrieved,
                )),
                error => Err(err_fmod!("FMOD_Studio_Bank_GetPath", error)),
            }
        }
    }
```
This should be converted to something like this: (matching the C# API pretty much)
```rs
    // Please excuse my crude indentation -_-
    pub fn get_path(&self) -> Result<String, Error> {
        unsafe {
            let mut retrieved = i32::default();
            match ffi::FMOD_Studio_Bank_GetPath(self.pointer, std::ptr::null(), 0, &mut retrieved)
            {
                ffi::FMOD_OK => Ok("".to_string()),
                ffi::FMOD_ERR_TRUNCATED => {
                   let path = CString::from_vec_unchecked(vec![b""; retrieved as usize]).into_raw();
                   match ffi::FMOD_Studio_Bank_GetPath(self.pointer, path, retrieved, &mut retrieved) {
                      ffi::FMOD_OK => Ok(CString::from_raw(path)
                        .into_string()
                        .map_err(Error::String)?,
                      ),
                      error => Err(err_fmod!("FMOD_Studio_Bank_GetPath", error)),
                   }
                }
                error => Err(err_fmod!("FMOD_Studio_Bank_GetPath", error)),
            }
        }
    }
```

It may be a good idea to parse the C# bindings as they more closely rusty style bindings or instead to use libfmod-gen to bootstrap most of the boilerplate code and manually edit the bindings from there.

**Comment by lebedec** [comment]
Yep, in some cases FMOD C headers do not provide enough information about parameter types. 
It is impossible to understand automatically where the pointer is used as an out parameter and where is an array of values or bytes buffer.
I parse HTML documentation to complete information about out parameters, but there are errors too.

All this leads to incorrect types and parameters.

I will fix functions in your examples.

And I think you just gave me a good idea - use C# code as more reliable source about types and out parameters instead of HTML documentation

**Comment by lebedec** [comment]
So, I made a mechanism for rewriting functions manually to deal with situations like this.
I think to switch manual edit the binding in future when these rewrites become too much and automatic generation will will make no sense.

Fixes for your case (update libfmod to 2.2.604):

```rust
let mut file = File::open("./data/SFX.bank")?;
let mut buffer = Vec::new();
file.read_to_end(&mut buffer)?;

let bank = studio.load_bank_memory(&buffer, FMOD_STUDIO_LOAD_BANK_NORMAL)?;
let path = bank.get_path()?;
let is_valid = bank.is_valid();
```

Thanks for help!

**Comment by melody-rs** [comment]
Tysm!!
I'm actually writing up fmod bindings to ruby atm which means I'm going over all of the functions in FMOD. I've been manually writing up fixed functions, I'd be more than happy to pull request them over when I'm finished!
